File: 20230415_Saturday/20230415.rb
# Leetcode: 2218. Maximum Value of K Coins From Piles.
# https://leetcode.com/problems/maximum-value-of-k-coins-from-piles/
# @param {Integer[][]} piles
# @param {Integer} k
# @return {Integer}
# 2023.04.15 Daily Challenge.
# TLE.
def max_value_of_coins(piles, k)
  # Based on:
  # https://leetcode.com/problems/maximum-value-of-k-coins-from-piles/solution/
  n = piles.size
  dp = Array.new(n+1).map{Array.new(k+1,0)}
  (1..n).each do |i|
    (0..k).each do |coins|
      sum = 0
      (0..[piles[i-1].size,coins].min).each do |cc| # Current coins.
        sum += piles[i-1][cc-1] if cc > 0
        dp[i][coins] = [dp[i][coins],dp[i-1][coins-cc]+sum].max
      end
    end
  end
  return dp[n][k]
end

File: 20230415_Saturday/20230415.cpp
// Leetcode: 2218. Maximum Value of K Coins From Piles.
// https://leetcode.com/problems/maximum-value-of-k-coins-from-piles/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 612 ms, faster than 12.82% of C++ online submissions for Maximum
// Value of K Coins From Piles.
// Memory Usage: 18.2 MB, less than 65.97% of C++ online submissions for Maximum
// Value of K Coins From Piles.
// 2023.04.15 Daily Challenge.
class Solution {
public:
    int maxValueOfCoins(vector<vector<int>>& piles, int k) {
      // Copied from:
      // https://leetcode.com/problems/maximum-value-of-k-coins-from-piles/solution/
        int n = piles.size();
        vector dp(n + 1, vector<int>(k + 1));
        for (int i = 1; i <= n; i++) {
            for (int coins = 0; coins <= k; coins++) {
                int currentSum = 0;
                for (int currentCoins = 0; currentCoins <= min((int)piles[i - 1].size(), coins); currentCoins++) {
                    if (currentCoins > 0) {
                        currentSum += piles[i - 1][currentCoins - 1];
                    }
                    dp[i][coins] = max(dp[i][coins], dp[i - 1][coins - currentCoins] + currentSum);
                }
            }
        }
        return dp[n][k];
    }
};
